src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -213,16 +225,41 @@ public static CallGraphEdge CombineEdges(CallGraphNode targetNode, params CallGr
     /// </summary>
     public class CallGraphNode : IEquatable<CallGraphNode>
     {
-        public readonly QsQualifiedName CallableName;
-        public readonly QsSpecializationKind Kind;
-        public readonly QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
+        // Outside of the Initialize method, these should be treated as read-only
+        public QsQualifiedName CallableName { get; private set; }
+        public QsSpecializationKind Kind { get; private set; }
+        public QsNullable<ImmutableArray<ResolvedType>> TypeArgs { get; private set; }
 
         /// <summary>
         /// Constructor for CallGraphNode objects.
         /// Strips position info from given type arguments before assigning them to TypeArgs.
         /// Throws an ArgumentNullException if callableName or kind is null.
         /// </summary>
         public CallGraphNode(QsQualifiedName callableName, QsSpecializationKind kind, QsNullable<ImmutableArray<ResolvedType>> typeArgs)
+            => Initialize(callableName, kind, typeArgs);
+
+        /// <summary>
+        /// Constructor for CallGraphNode objects.
+        /// Strips position info from given type arguments before assigning them to TypeArgs.
+        /// Throws an ArgumentNullException if specialization is null.
+        /// </summary>
+        public CallGraphNode(QsSpecialization specialization)
+        {
+            if (specialization == null) throw new ArgumentNullException(nameof(specialization));
+            Initialize(specialization.Parent, specialization.Kind, specialization.TypeArguments);
+        }
+
+        /// <summary>
+        /// Copy constructor for CallGraphNode objects.
+        /// Throws an ArgumentNullException if node is null.
+        /// </summary>
+        public CallGraphNode(CallGraphNode node)
```

Comment:
I think that the `private set` vs `readonly` doesn't affect whether copying is needed; as long as the private set is only used by the constructor then the class is effectively immutable. Although it is nicer if the fields aren't settable. I did have a couple other ideas for making the fields readonly without using expression `throw`s though:

* It looks like there is a way to do it by [passing the fields to `Initialize` as `ref` parameters](https://stackoverflow.com/a/6795985), but it looks kind of weird.
* Alternatively, if you `#nullable enable` in this file, you can remove the null argument checks and make `Initialize` an alternate constructor called with `: this(...)`.

Reply:
The code now uses the `this` syntax to consolodate constructor logic, rather than using an Initialize method. Using `readonly` keyword instead of `private set` syntax. @samarsha I will want to look more into the `#nullable enable` feature later to see if that can be used to clean up a lot of the null checks in the file. I will open a separate conversation on this PR to track that.

Reply:
@samarsha Here is the conversation tracking the use of the `#nullable enable` feature: https://github.com/microsoft/qsharp-compiler/pull/454#discussion_r439683548